### DISCUSSION TITLE
Hydration Error in Next.js 15 with React 19 Due to cz-shortcut-listen="true" Attribute Injection from Colorzilla Extension

### DISCUSSION THREAD

**Discussion by MonaAghili** [vercel/next.js#72035] [Help] [answered]
### To Reproduce

**Description:**

After updating my Next.js application to:

- **Next.js:** 15.0.2
- **React:** 19.0.0-rc-02c0e824-20241028
- **React-DOM:** 19.0.0-rc-02c0e824-20241028

I encountered a new hydration error that was not present in previous versions. Through investigation, I discovered that the `cz-shortcut-listen="true"` attribute, added by the Colorzilla browser extension, was causing the issue.

**Steps to reproduce:**

1. Update Next.js and React versions as mentioned above.
2. Enable the Colorzilla browser extension.
3. Run the Next.js application and inspect any hydration errors.

**Findings:**

The `cz-shortcut-listen="true"` attribute injected by Colorzilla causes a mismatch during the server-client hydration process in Next.js. When Colorzilla is disabled, the hydration error resolves. However, Colorzilla is a tool we rely on heavily, and disabling it whenever using Next.js is not practical.

### Current vs. Expected behavior

**Request:**

Could the Next.js team suggest a solution or workaround to prevent this attribute from interfering with hydration without disabling Colorzilla? This would be immensely helpful for developers who rely on both Next.js and the Colorzilla extension. Thank you!

![image](https://github.com/user-attachments/assets/1d993bd1-502c-4ec5-b80a-fd547497d3eb)

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
  Available memory (MB): 12168
  Available CPU cores: 8
Binaries:
  Node: 20.12.2
  npm: N/A
  Yarn: N/A
  pnpm: 9.4.0
Relevant Packages:
  next: 15.0.2 // Latest available version is detected (15.0.2).
  eslint-config-next: 14.1.3
  react: 19.0.0-rc-02c0e824-20241028
  react-dom: 19.0.0-rc-02c0e824-20241028
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local)

**Comment by evaldoprestes** [comment]
+1

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by ztanner** [comment]
Hi,

As these hydration errors come from React and not Next.js, this isn't actionable as a bug report. Going to convert this to a discussion in case other folks have advice related to the ColorZilla extension.

**Comment by bsaisushanth99999** [comment]
Chances are you are using **"colorzilla"** extention,

 cz-shortcut-listen is inserted into html by the colorzilla extension/add-on. Remove that and everything should work fine!

**Reply by MonaAghili** [reply]
I literally mentioned in my initial description that ColorZilla was causing this issue. So, are we really expected to disable all helpful extensions just to bypass it?

**Reply by felixbouazza** [reply]
Works for me, thanks @bsaisushanth99999

**Reply by SamWhillance** [reply]
Awesome, thanks for the fix!

**Reply by idoamram** [reply]
Thanks a lot!

**Reply by BodaNabeel** [reply]
Thanks worked for me.

**Reply by salahchaabi** [reply]
@bsaisushanth99999 thank you very much!
It helped a lot!

**Reply by maxschneidercodes** [reply]
Thanks

**Reply by RadmanRAZ** [reply]
worked !!! thanks a lot

**Reply by petter-tchirkov** [reply]
that's the correct answer, much appreciated

**Comment by MasterProvider** [comment]
![Screenshot 2024-10-30 091658](https://github.com/user-attachments/assets/2b7cf78c-8a1b-4fea-a948-9d8297893139)

![Screenshot 2024-10-30 205403](https://github.com/user-attachments/assets/58150433-d57f-4795-9dc6-bfd560c311ce)
same issue even I did not use chrome extensions

**Reply by MonaAghili** [reply]
@MasterProvider 
Try using incognito mode in your browser to ensure no other factors, like cached data or extensions, are affecting the application. If the issue still occurs, there may be a problem in the source code causing a hydration mismatch. Double-checking your server-side and client-side rendering setup might help!

**Reply by MasterProvider** [reply]
I already test it before I raised my concern

**Reply by MonaAghili** [reply]
I understand this issue is different from what I’ve encountered. If you find a solution, please feel free to share it, having another approach on hand could be really helpful if similar issues come up in the future. Thanks, and good luck with troubleshooting!

**Reply by icyJoseph** [reply]
This one is most likely because MUI or whatever is not being told that it is running in SSR, so it generated non-deterministic classnames

**Reply by MasterProvider** [reply]
> ![Screenshot 2024-10-30 091658](https://private-user-images.githubusercontent.com/48547307/381559500-2b7cf78c-8a1b-4fea-a948-9d8297893139.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMTE4MzgsIm5iZiI6MTczNTExMTUzOCwicGF0aCI6Ii80ODU0NzMwNy8zODE1NTk1MDAtMmI3Y2Y3OGMtOGExYi00ZmVhLWE5NDgtOWQ4Mjk3ODkzMTM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDA3MjUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkMzQxYWNhN2VlZDNlY2E1Yzg1OGNlOTA5MGZkNmY5YmI2NjdmZWIwNDAwZjdmZGVjNDNmM2JhNTk3OTJjNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZctU1HivYGmGmp-IFose7l532zHCiyOw2S0fPlGaZZw)
> 
> ![Screenshot 2024-10-30 205403](https://private-user-images.githubusercontent.com/48547307/381558969-58150433-d57f-4795-9dc6-bfd560c311ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMTE4MzgsIm5iZiI6MTczNTExMTUzOCwicGF0aCI6Ii80ODU0NzMwNy8zODE1NTg5NjktNTgxNTA0MzMtZDU3Zi00Nzk1LTlkYzYtYmZkNTYwYzMxMWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDA3MjUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0MmQ4ODM0ZmUwNjJjYWE2NTQ5ODc0NmUyY2MyMTllYzE1ZDQzN2YyZWE0ZjQwMWIxZWQyMmZjZmNlYzFmYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8Vg2MrtzglGX8BnjT93Vv-KXIy726J3Jgx-rd6QtVpE) same issue even I did not use chrome extensions

This issue was caused by a hydration problem in TSS-React. It no longer supports nested class selector in Next.js 15

**hydration error**

```
sideViewContent:{
    height:'auto',
    position:'relative',
    padding: '6.4rem',
    [theme.breakpoints.down('lg')]: {
      padding: '2.4rem',
    },
    [`& .${classes.appIcon}`]: {
      height: '70px',
      width: '70px'
    },
}
```
**Fixed version** 
```
sideViewContent:{
    height:'auto',
    position:'relative',
    padding: '6.4rem',
    [theme.breakpoints.down('lg')]: {
      padding: '2.4rem',
    }
},
appIcon: {
      height: '70px',
      width: '70px'
},
```
**or do it like this**
```
sideViewContent:{
    height:'auto',
    position:'relative',
    padding: '6.4rem',
    [theme.breakpoints.down('lg')]: {
      padding: '2.4rem',
    },
    '& .app-icon': {
      height: '70px',
      width: '70px'
   },
}
```

**Comment by icyJoseph** [comment]
According to, https://react.dev/blog/2024/04/25/react-19#compatibility-with-third-party-scripts-and-extensions, hydration errors could be skipped if the changed tags are the body, or head... 

On which tag is this happening?

**Reply by MonaAghili** [reply]
the body tag

**Reply by icyJoseph** [reply, marked as answer]
Could you add `suppressHydrationWarning` to the body tag? I guess you are in control of that.

Intriguing that it doesn't work though. Maybe something to report to React? I don't want to speculate, but maybe having two mismatches triggers the warning anyway?

**Reply by MasterProvider** [reply]
thank for suggestion I tried it but suppressHydrationWarning did not suppress the warning

**Reply by icyJoseph** [reply]
@MasterProvider not for your case. In your case the class names generated in server and client are diverging, that usually happens when the CSS in JS library is not setup for SSR

**Reply by MonaAghili** [reply]
> Could you add `suppressHydrationWarning` to the body tag? I guess you are in control of that.
> 
> Intriguing that it doesn't work though. Maybe something to report to React? I don't want to speculate, but maybe having two mismatches triggers the warning anyway?

Thank you so much @icyJoseph! I’ll give it a try to see if it solves the problem

**Reply by MonaAghili** [reply]
> Could you add `suppressHydrationWarning` to the body tag? I guess you are in control of that.
> 
> Intriguing that it doesn't work though. Maybe something to report to React? I don't want to speculate, but maybe having two mismatches triggers the warning anyway?

This solution resolved the issue.
Thank you

**Reply by icyJoseph** [reply]
Great! Do remember that supressHydrationWarning:

> This only works one level deep, and is intended to be an escape hatch. Don’t overuse it. Unless it’s text content, React still won’t attempt to patch it up, so it may remain inconsistent until future updates.

- https://react.dev/reference/react-dom/hydrate#suppressing-unavoidable-hydration-mismatch-errors

**Reply by batchor** [reply]
> Could you add `suppressHydrationWarning` to the body tag? I guess you are in control of that.
> 
> Intriguing that it doesn't work though. Maybe something to report to React? I don't want to speculate, but maybe having two mismatches triggers the warning anyway?

I encounter this error on the html tag. Should I add `suppressHydrationWarning` to html tag also?

**Reply by icyJoseph** [reply]
Give it a go. And report back.

**Reply by batchor** [reply]
> Give it a go. And report back.

I added `suppressHydrationWarning` to html tag and the error goes away. But I'm paranoid that if there is any side effects or performance impact.

**Reply by icyJoseph** [reply]
Only that, if you were to introduce an actual hydration error in the HTML tag, you would be suppressing it... But that's not gonna be the case. 

The supressHydrationWarning prop works only at that very same tag level, so issues down the tree will not be suppressed.

It might actually be necessary, because if your users have that extension installed...

**Reply by YuHyeonWook** [reply]
Thank you!

**Reply by jacktsin1** [reply]
@icyJoseph will that prevent an RSC from being rendered on the client as the error suggests, or will it just suppress the error message?

**Comment by b3nThomas** [comment]
As an alternative, you can tweak the settings to only allow CZ to modify the page when you actually click the extension.
So you're only opting in to the hydration errors for the lifespan of that tab when you're actually using the color picker.
I personally much prefer this to suppressing hydration errors in any way in React.

<img width="564" alt="Screenshot 2025-01-06 at 20 45 50" src="https://github.com/user-attachments/assets/32f8f4dc-0653-4ba8-9956-c8f61047baae" />

**Reply by beamcode** [reply]
This was the issue for me thanks genius.

**Reply by wutangpaul** [reply]
Great solution, thank you

**Reply by KyorCode** [reply]
This is the by far the best solution.

**Comment by bloodykheeng** [comment]
for me ive tried all solutions and the have failed what i had to do was to disable the browser extension instead and i stop getting the hydration error

hear is the error i got
```
Uncaught Error: Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used:

- A server/client branch if (typeof window !== 'undefined').
- Variable input such as Date.now() or Math.random() which changes each time it's called.
- Date formatting in a user's locale which doesn't match the server.
- External changing data without sending a snapshot of it along with the HTML.
- Invalid HTML tag nesting.

It can also happen if the client has a browser extension installed which messes with the HTML before React loaded.                              <RootLayout>
                            <div
+                             className="surface-0 flex justify-content-center"
-                             className={null}
-                             id="scrnli_recorder_root"
                            >
+                             <div id="home" className="landing-wrapper overflow-hidden">
```



i noted that i dont  have this 
```
    className={null}
-                             id="scrnli_recorder_root"
```

in my code anywhere so it was a browser extension injecting it instead so i to go and disable the scrnly brwser extension
Screenshot & Screen Recorder Tool  

[Screenshot & Screen Recorder Tool ](https://chromewebstore.google.com/detail/screenshot-screen-recorde/ijejnggjjphlenbhmjhhgcdpehhacaal?pli=1)


and now my code is smooth

**Comment by kosiakMD** [comment]
the same issue (related to Grammarly), that's creepy

**Comment by foysal-m** [comment]
I get the following error, even though I tried removing all extensions, and suppressing the error, but it just does not work. Tried with incognito mode too. It's. very bad, as it prevents to write cypress tests as it shows the same error for cypress headless browser

<img width="644" alt="image" src="https://github.com/user-attachments/assets/83e24da5-6f1f-4fb5-b84b-c61fe5314bc7" />

**Reply by bloodykheeng** [reply]
Put tanstack in a separate file with use client then wrap it over your main app

**Reply by foysal-m** [reply]
Thanks for looking into this, the problem was using a html webpack plugin which generates a new html on every build. Then I solved it, not having any html from the client side, rather passing all from the server side. Then there is no issues anymore.Hope this makes sense

**Reply by bloodykheeng** [reply]
Cool 😎

**Comment by naglaa77** [comment]
✅ Solution (Workaround)
The hydration error is caused by the cz-shortcut-listen="true" attribute injected by the ColorZilla browser extension before React hydration begins. Since this attribute isn’t present in the server-rendered HTML, React throws a hydration mismatch warning in strict mode.

To fix this without disabling the extension, I injected a small client-side script that:

- Removes the attribute before hydration

- Uses a MutationObserver to prevent it from being re-added
 ### Implementation
components/HydrationFix.tsx
```

`'use client';
import React from 'react';

export function HydrationFix(): React.JSX.Element {
  return (
    <script
      id="hydration-fix"
      dangerouslySetInnerHTML={{
        __html: `
          if (document.body.hasAttribute('cz-shortcut-listen')) {
            document.body.removeAttribute('cz-shortcut-listen');
          }

          const observer = new MutationObserver((mutations) => {
            mutations.forEach((mutation) => {
              if (
                mutation.type === 'attributes' &&
                mutation.attributeName === 'cz-shortcut-listen'
              ) {
                document.body.removeAttribute('cz-shortcut-listen');
              }
            });
          });

          observer.observe(document.body, {
            attributes: true,
            attributeFilter: ['cz-shortcut-listen']
          });
        `,
      }}
    />
  );
}
`

components/ClientLayout.tsx
`'use client';
import { HydrationFix } from './HydrationFix';

export default function ClientLayout({ children }: { children: React.ReactNode }) {
  return (
    <>
      <HydrationFix />
      {children}
    </>
  );
}
`
app/layout.tsx
`
import ClientLayout from '@/components/ClientLayout';

export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (
    <html lang="en">
      <body>
        <ClientLayout>{children}</ClientLayout>
      </body>
    </html>
  );
}
`React compares the DOM generated by the server with what exists in the browser before hydration. If browser extensions like ColorZilla modify the DOM in between, React detects a mismatch.

This script removes the unwanted attribute before hydration, ensuring React sees a matching DOM. The MutationObserver prevents the attribute from coming back.

**Comment by Mharbulous** [comment]
I was having this error, but a bit more serious because my Claude Chats were getting stuck in an unrecoverable state of "Starting Claude Code..."  forever.  

In my case, after a little winnowing, it turned out to be the Vue.js chrome extension that was causing hydration errors which sometimes led to getting stuck in an unrecoverable state.
